File: lib/incrementalist/game/big_num.ex
defmodule BigNum do
  @moduledoc """
  High-precision large number representation (m * 10^e).
  Maintains 15 digits of precision.
  """

  use Ecto.Schema
  import Ecto.Changeset

  @precision_digits 15
  @add_cutoff @precision_digits + 1

  @primary_key false
  @derive Jason.Encoder
  embedded_schema do
    field :m, :float, default: 0.0
    field :e, :integer, default: 0
  end

  def changeset(schema \\ %__MODULE__{}, attrs) do
    schema
    |> cast(attrs, [:m, :e])
    |> validate_required([:m, :e])
  end

  def zero, do: %BigNum{m: 0.0, e: 0}
  def one, do: %BigNum{m: 1.0, e: 0}

  @doc """
  Normalizes a BigNum so that 1 <= abs(m) < 10.
  """
  def normalize(%BigNum{m: m}) when m == 0.0, do: %BigNum{m: 0.0, e: 0}

  def normalize(%BigNum{m: m, e: e}) do
    shift = Float.floor(:math.log10(abs(m))) |> trunc()

    m = m / :math.pow(10, shift)
    e = e + shift

    {m, e} =
      cond do
        abs(m) >= 10 ->
          {m / 10, e + 1}

        abs(m) < 1 ->
          {m * 10, e - 1}

        true ->
          {m, e}
      end

    # Keep roughly 15 significant digits
    m =
      m
      |> :erlang.float_to_binary(decimals: @precision_digits - 1)
      |> String.to_float()

    {m, e} =
      if abs(m) >= 10 do
        {m / 10, e + 1}
      else
        {m, e}
      end

    %BigNum{m: m, e: e}
  end

  def from_number(v) when v == 0, do: zero()

  def from_number(value) when is_number(value) do
    normalize(%BigNum{m: value * 1.0, e: 0})
  end

  def new(m, e) do
    normalize(%BigNum{m: m * 1.0, e: e})
  end

  @doc """
  Converts a BigNum to a native float.
  Warning: May return :inf or 0.0 if outside double-precision range.
  """
  def to_float(%BigNum{m: m, e: e}) do
    m * :math.pow(10, e)
  end

  def mul(%BigNum{m: m}, _b) when m == 0.0, do: zero()
  def mul(_a, %BigNum{m: m}) when m == 0.0, do: zero()

  def mul(%BigNum{} = a, %BigNum{} = b) do
    normalize(%BigNum{
      m: a.m * b.m,
      e: a.e + b.e
    })
  end

  def div(_a, %BigNum{m: m}) when m == 0.0 do
    raise ArithmeticError, message: "division by zero"
  end

  def div(%BigNum{m: m}, _b) when m == 0.0, do: zero()

  def div(%BigNum{} = a, %BigNum{} = b) do
    normalize(%BigNum{
      m: a.m / b.m,
      e: a.e - b.e
    })
  end

  def add(%BigNum{m: m}, b) when m == 0.0, do: b
  def add(a, %BigNum{m: m}) when m == 0.0, do: a

  def add(%BigNum{} = a, %BigNum{} = b) do
    diff = a.e - b.e

    cond do
      diff > @add_cutoff ->
        a

      diff < -@add_cutoff ->
        b

      true ->
        {upper, lower, abs_diff} =
          if diff < 0 do
            {b, a, -diff}
          else
            {a, b, diff}
          end

        normalize(%BigNum{
          m: upper.m + lower.m * :math.pow(10, -abs_diff),
          e: upper.e
        })
    end
  end

  def sub(%BigNum{m: m}, %BigNum{} = b) when m == 0.0, do: %BigNum{m: -b.m, e: b.e}
  def sub(%BigNum{} = a, %BigNum{m: m}) when m == 0.0, do: a

  def sub(%BigNum{} = a, %BigNum{} = b) do
    diff = a.e - b.e

    cond do
      diff > @add_cutoff -> a
      diff < -@add_cutoff -> %BigNum{m: -b.m, e: b.e}
      true -> add(a, %BigNum{m: -b.m, e: b.e})
    end
  end

  def pow(%BigNum{m: m}, _p) when m == 0.0, do: zero()

  def pow(%BigNum{} = a, p) when is_number(p) do
    if a.m < 0 and p != trunc(p) do
      raise ArithmeticError, message: "cannot raise negative number to non-integer power"
    end

    log10_value = :math.log10(abs(a.m)) + a.e
    result_log = log10_value * p

    e = Float.floor(result_log) |> trunc()
    m = :math.pow(10, result_log - e)

    m =
      if a.m < 0 and rem(trunc(p), 2) != 0 do
        -m
      else
        m
      end

    normalize(%BigNum{m: m, e: e})
  end

  def compare(%BigNum{m: m1}, %BigNum{m: m2}) when m1 == 0.0 and m2 == 0.0, do: 0
  def compare(%BigNum{m: m1}, %BigNum{m: m2}) when m1 >= 0 and m2 < 0, do: 1
  def compare(%BigNum{m: m1}, %BigNum{m: m2}) when m1 < 0 and m2 >= 0, do: -1

  def compare(%BigNum{} = a, %BigNum{} = b) do
    sign = if a.m < 0, do: -1, else: 1

    cond do
      a.e > b.e -> sign
      a.e < b.e -> -sign
      abs(a.m) > abs(b.m) -> sign
      abs(a.m) < abs(b.m) -> -sign
      true -> 0
    end
  end

  def format(%BigNum{m: m}) when m == 0.0, do: "0"

  def format(%BigNum{} = n) do
    mantissa = :erlang.float_to_binary(n.m, decimals: @precision_digits - 1)
    "#{mantissa}e#{n.e}"
  end
end

File: assets/src/core/bignum.ts
export type BigNum = {
  m: number // 15-digit mantissa, normalized: 1 <= abs(m) < 10
  e: number // base-10 exponent
}

export const PRECISION_DIGITS = 15
export const ADD_CUTOFF = PRECISION_DIGITS + 1

export const ZERO: BigNum = { m: 0, e: 0 }
export const ONE: BigNum = { m: 1, e: 0 }

/**
 * Normalizes a BigNum so that 1 <= abs(m) < 10.
 */
export function normalize(n: BigNum): BigNum {
  if (n.m === 0 || !Number.isFinite(n.m)) {
    return { m: 0, e: 0 }
  }

  const shift = Math.floor(Math.log10(Math.abs(n.m)))

  let m = n.m / 10 ** shift
  let e = n.e + shift

  // Guard against floating edge cases like 9.999999999999999 becoming 10
  if (Math.abs(m) >= 10) {
    m /= 10
    e += 1
  }

  if (Math.abs(m) < 1) {
    m *= 10
    e -= 1
  }

  // Keep roughly 15 significant digits
  m = Number(m.toPrecision(PRECISION_DIGITS))

  if (Math.abs(m) >= 10) {
    m /= 10
    e += 1
  }

  return { m, e }
}

export function fromNumber(value: number): BigNum {
  if (value === 0) return ZERO
  return normalize({ m: value, e: 0 })
}

export function big(m: number, e: number): BigNum {
  return normalize({ m, e })
}

/**
 * Converts a BigNum to a native number. 
 * Warning: May return Infinity or 0 if outside double-precision range.
 */
export function toNumber(n: BigNum): number {
  return n.m * 10 ** n.e
}

export function mul(a: BigNum, b: BigNum): BigNum {
  if (a.m === 0 || b.m === 0) return ZERO

  return normalize({
    m: a.m * b.m,
    e: a.e + b.e,
  })
}

export function div(a: BigNum, b: BigNum): BigNum {
  if (b.m === 0) throw new Error("Division by zero")
  if (a.m === 0) return ZERO

  return normalize({
    m: a.m / b.m,
    e: a.e - b.e,
  })
}

export function add(a: BigNum, b: BigNum): BigNum {
  if (a.m === 0) return b
  if (b.m === 0) return a

  const diff = a.e - b.e
  if (diff > ADD_CUTOFF) return a
  if (diff < -ADD_CUTOFF) return b

  // Make a the larger exponent
  let upper = a
  let lower = b
  let absDiff = diff
  if (diff < 0) {
    upper = b
    lower = a
    absDiff = -diff
  }

  return normalize({
    m: upper.m + lower.m * 10 ** -absDiff,
    e: upper.e,
  })
}

export function sub(a: BigNum, b: BigNum): BigNum {
  if (a.m === 0) return { m: -b.m, e: b.e }
  if (b.m === 0) return a

  const diff = a.e - b.e
  if (diff > ADD_CUTOFF) return a
  if (diff < -ADD_CUTOFF) return { m: -b.m, e: b.e }

  return add(a, { m: -b.m, e: b.e })
}

/**
 * Power by normal number. Good for game formulas like x^1.15, x^2, etc.
 */
export function pow(a: BigNum, p: number): BigNum {
  if (a.m === 0) return ZERO
  if (a.m < 0 && !Number.isInteger(p)) {
    throw new Error("Cannot raise negative number to non-integer power")
  }

  const log10Value = Math.log10(Math.abs(a.m)) + a.e
  const resultLog = log10Value * p

  const e = Math.floor(resultLog)
  let m = 10 ** (resultLog - e)

  if (a.m < 0 && Number.isInteger(p) && p % 2 !== 0) {
    m = -m
  }

  return normalize({ m, e })
}

export function compare(a: BigNum, b: BigNum): number {
  if (a.m === 0 && b.m === 0) return 0
  if (a.m >= 0 && b.m < 0) return 1
  if (a.m < 0 && b.m >= 0) return -1

  const sign = a.m < 0 ? -1 : 1

  if (a.e > b.e) return sign
  if (a.e < b.e) return -sign
  if (Math.abs(a.m) > Math.abs(b.m)) return sign
  if (Math.abs(a.m) < Math.abs(b.m)) return -sign
  return 0
}

export function formatScientific(n: BigNum, digits = 15): string {
  if (n.m === 0) return "0"
  return `${n.m.toPrecision(digits)}e${n.e}`
}
